src/com/sap/piper/k8s/ContainerMap.groovy

```diff
@@ -19,4 +20,78 @@ class ContainerMap implements Serializable {
     void setMap(Map containersMap) {
         containerMap = containersMap
     }
+
+    void initFromResource(Script script, String yamlResourceName, String buildTool) {
+        script.echo "initFromResource(yamlResourceName: $yamlResourceName, buildTool: $buildTool)"
+        Map containers = [:]
+        boolean[] piperExecutionPrepared = new boolean[1]
+        try {
+            Map yamlContents = script.readYaml(text: script.libraryResource(yamlResourceName))
+            Map stageToStepMapping = yamlContents.containerMaps as Map
+            Map stepToMetaDataMapping = yamlContents.stepMetadata as Map ?: [:]
+            stageToStepMapping.each { stageName, stepsList ->
+                containers[stageName] = getContainersForStage(script, stageName as String, stepsList as List,
+                    stepToMetaDataMapping, buildTool, piperExecutionPrepared)
+            }
+        } catch (Exception e) {
+            script.error "Failed to parse container maps in '$yamlResourceName'. It is expected to contain " +
+                "the entries 'containerMaps' and optionally 'stepMetadata' in the root." +
+                "containerMaps which is a map of stage identifiers to a list of executed steps. " +
+                "The optional 'stepMetadata' is a map of (go-implemented) step names to their YAML " +
+                "metadata resource file." +
+                "Error: ${e.getMessage()}"
+        }
+        setMap(containers)
+    }
+
+    static Map getContainersForStage(Script script, String stageName, List stepsList, Map stepToMetaDataMapping, String buildTool, boolean[] piperExecutionPrepared) {
+        Map containers = [:]
+        stepsList.each { stepName ->
+            String imageName = getDockerImageNameForStepInStage(script, stageName, stepName as String, buildTool)
```

Comment:
The only risk I see is that something is defined as piper default. However, the defaults from piper are not used later. If they are different from the metadata file you will have a wrong image. But I think we do not have to consider that.

Reply:
> The only risk I see is that something is defined as piper default. However, the defaults from piper are not used later. If they are different from the metadata file you will have a wrong image. But I think we do not have to consider that.

You are correct. Looking into default_pipeline_environment.yml, there are still many defaults for steps which are now go. For example mavenExecute. I can just revert to a version where go steps were always resolved using piper.

Reply:
For our use-case the performance impact is about 15 to 20 seconds more for resolving all Go steps purely using the piper binary.